Mu2eKinKal/src/LoopHelixFit_module.cc

```diff
@@ -347,88 +475,86 @@ namespace mu2e {
-          if(helix.radius() == 0.0 || helix.lambda() == 0.0 )
-            throw cet::exception("RECO")<<"mu2e::HelixFit: degenerate seed parameters" << endl;
-          auto zcent = Mu2eKinKal::zMid(hseed.hits());
-          // take the magnetic field at the helix center as nominal
-          VEC3 center(helix.centerx(), helix.centery(),zcent);
-          auto bnom = kkbf_->fieldVect(center);
-          // compute the charge from the helicity, fit direction, and BField direction
-          double bz = bnom.Z();
-          int charge = static_cast<int>(copysign(PDGcharge_,(-1)*helix.helicity().value()*fdir_.dzdt()*bz));
-          // test consistency.  Modify this later when the HelixSeed knows which direction it's going TODO
-          auto fitpart = fpart_;
-          if(charge*PDGcharge_ < 0){
-            if(usePDGCharge_)throw cet::exception("RECO")<<"mu2e::HelixFit: inconsistent charge" << endl;
-            fitpart = static_cast<PDGCode::type>(-1*fitpart); // reverse sign
+
+        // determine the fit direction hypothesis
+        auto helix_dir = chooseHelixDir(hseed);
+        const bool undefined_dir = helix_dir == HelixRecoDir::PropDir::ambiguous; //check if no direction is significantly preferred
+
+        // fit the track hypothesis (hypotheses)
+        std::unique_ptr<KKTRK> ktrk = nullptr;
+        if(undefined_dir) {
+          ++nAmbiguous_;
+          auto ktrk_down = fitTrack(event, hseed,  1.f, fpart_);
+          auto ktrk_up   = fitTrack(event, hseed, -1.f, fpart_);
+          const bool use_downstream = !ktrk_up || (ktrk_down && compare_tracks(*ktrk_down, *ktrk_up)); //if one is null, default to the other (use down if both are null, doesn't matter)
+          if(print_ > 0)
+            printf("[LoopHelixFit::%s]: Fit both direction options, slope = %9.2e, sig = %.2f, status down = %i, status up = %i, p(chi^2) down = %.4f, p(chi^2) up = %.4f: chose %s\n",
+                   __func__, hseed.recoDir().slope(), hseed.recoDir().slopeSig(),
+                   ktrk_down && goodFit(*ktrk_down), ktrk_up && goodFit(*ktrk_up),
+                   (ktrk_down) ? ktrk_down->fitStatus().chisq_.probability() : -1.,
+                   (ktrk_up  ) ? ktrk_up  ->fitStatus().chisq_.probability() : -1.,
+                   (use_downstream) ? "downstream" : "upstream");
+          ktrk = (use_downstream) ? std::move(ktrk_down) : std::move(ktrk_up);
+          helix_dir = (use_downstream) ? HelixRecoDir::PropDir::downstream : HelixRecoDir::PropDir::upstream; //record which direction was chosen
+        } else {
+          ktrk = fitTrack(event, hseed, helix_dir, fpart_);
+        }
+
+        if(!ktrk) continue; //ensure that the track exists
+
+        // Check the fit
+        auto goodfit = goodFit(*ktrk);
+
+        // extrapolate as required
+        if(goodfit && extrapolate_) extrapolate(*ktrk);
+        if(print_>1) ktrk->printFit(std::cout,print_-1);
+
+        // save the fit result
+        if(goodfit || saveall_){
```

Comment:
since goodfit has been tested several times already before getting to this statement I think the saveall_ flag no longer does anything. We should discuss whether to remove it or redefine it.

Reply:
I think with `saveall_` set it will still save the track bad fit track (possibly without the drift fit or extrapolation performed) as was true before. In the case both fits are performed and both are bad, the downstream is saved with this flag. I don't know what this feature is used for, so I'm open to removing or redefining it in anyway that is suggested!

Reply:
saveall_ is for debugging failed fits. It's never used in production.